_posts/2020-02-06-how-to-configure-a-production-grade-ci-cd-setup-for-apps-and-infrastructure-code.adoc

```diff
@@ -0,0 +1,2045 @@
+  model, having passed numerous compliance audits including FedRamp. They provide a wide variety of features to mitigate
+  potential threats, including runtime environment isolation, restricted contexts for finer grained permissions
+  modeling, and audit logging to continuously monitor access.
+
+https://about.gitlab.com/[GitLab] (self-hosted and Saas, generic)::
+  GitLab is an all comprehensive platform that supports both a self-hosted mode (enterprise) or Saas platform. Since
+  GitLab also provides hosting for git repositories, it creates a tight integration between the git workflows and the
+  CI/CD workflows that you can implement on the platform. Although GitLab is a generic CI/CD platform that can be
+  configured to run any workflow, GitLab also provides specialized workflows for popular infrastructure platforms like
+  Kubernetes. Some of its continuous deployment features include first class support for feature toggles, canary
+  deployments, and building docker images. While the Saas platform is subject to external attacks, GitLab employs a
+  respectable security model with end to end encryption, frequent testing, and compliance audits, in addition to being
+  an open source and open core platform that allows for wider inspection and review of its practices. For those that can
+  not rely on the Saas platform for CI/CD, you can always run GitLab EE in your own data center.
+
+https://www.runatlantis.io/[Atlantis] (self-hosted, specialized)::
+  Atlantis is an open source tool optimized for git based Terraform workflows, with additional support for Terragrunt.
+  At its core, the platform will automatically run `terraform plan` on commits and annotate any open Pull Requests with
+  the `plan` output. This `plan` is stored in plan file format so that when upon merge, the exact plan is applied using
+  `terraform apply`. Note that Atlantis is optimized for this single workflow. That is, it cannot be made to implement
+  other workflows, such as building images or running infrastructure tests using terratest. Being a self-hosted
+  platform, you can keep your credentials internal to your data center and lock down server level access. However,
+  Atlantis requires a public facing endpoint to be available so that it can listen on webhooks from the major VCS
+  platforms (GitHub/GitLab/Bitbucket/Azure DevOps). This means that you are relying on the security model of the
+  Atlantis platform to ensure that it doesn't accidentally run infrastructure code from unwanted sources with the
+  credentials of the platform. To prevent unwanted deployments, you will want to ensure that you are following
+  https://www.runatlantis.io/docs/security.html#mitigations[all the security best practices] of the platform.
+
+https://www.hashicorp.com/products/terraform/[Terraform Enterprise and Terraform Cloud] (self-hosted and Saas, specialized)::
```

Comment:
Mention somewhere that due to the limitations of Atlantis, TFC, and TFE, they can't be used as part of the secure CI / CD setup in this guide. 

Also, consider for each platform listing some of the trade-offs from: https://github.com/gruntwork-io/team-product/blob/master/prds/minimal-terraform-pipeline.md#why-build-it. In particular, whether it supports approval flows, commit triggers, manual triggers, secure permissions handling, etc.

In fact, maybe the way to simplify this section is to have a table. The columns are each platform and the rows are the various requirements. We can spell out what we mean by each requirement before the table and then present the summary in a simple to scan format that runs less risk of being "judgmental" of the platforms? And then we can say that the CI / CD platform we recommend for app and infra code should meet requirements A, B, and C, which you can see are met by these platforms but not these other ones.